app/uk/gov/hmrc/tariffclassificationfrontend/controllers/AttachmentsController.scala

```diff
@@ -0,0 +1,128 @@
+import uk.gov.hmrc.tariffclassificationfrontend.views.CaseDetailPage
+import uk.gov.hmrc.tariffclassificationfrontend.views.CaseDetailPage.CaseDetailPage
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.concurrent.Future.successful
+
+@Singleton
+class AttachmentsController @Inject()(authenticatedAction: AuthenticatedAction,
+                                      casesService: CasesService,
+                                      fileService: FileStoreService,
+                                      val messagesApi: MessagesApi,
+                                      implicit val appConfig: AppConfig,
+                                      implicit val mat: Materializer) extends FrontendController with I18nSupport {
+
+  private lazy val form: Form[UploadAttachmentFormData] = UploadAttachmentFormData.form
+  private val maxSizeMB = 100 * 1024 * 1024
+
+  def attachmentsDetails(reference: String): Action[AnyContent] = authenticatedAction.async { implicit request =>
+    getCaseAndRenderView(reference, CaseDetailPage.ATTACHMENTS, c => renderView(c, form))
+  }
+
+  def uploadAttachment(reference: String): Action[Either[MaxSizeExceeded, MultipartFormData[TemporaryFile]]] =
+    authenticatedAction.async(parse.maxLength(maxSizeMB, parse.multipartFormData)) { implicit request =>
+
+      request.body match {
+        case Left(MaxSizeExceeded(_)) => renderErrors(reference, Some(messagesApi("cases.attachment.upload.error.restrictionSize")))
+        case Right(multipartForm) => {
+          multipartForm match {
+            case file: MultipartFormData[TemporaryFile] if (!file.files.isEmpty) => uploadAndSave(reference, file)
+            case _ => renderErrors(reference, Some(messagesApi("cases.attachment.upload.error.mustSelect")))
+          }
+        }
+      }
+    }
+
+  private def uploadAndSave(reference: String, multiPartFormData: MultipartFormData[TemporaryFile])
+                           (implicit hc: HeaderCarrier, request: AuthenticatedRequest[_]) = {
+
+    multiPartFormData.file("file-input") match {
+      case Some(filePart) if filePart.filename.isEmpty => renderErrors(reference, Some(messagesApi("cases.attachment.upload.error.mustSelect")))
+      case Some(filePart) => {
+        val fileUpload = FileUpload(filePart.ref, filePart.filename, filePart.contentType.getOrElse(throw new IllegalArgumentException("Missing file type")))
+        casesService.getOne(reference).flatMap {
+          case Some(c: Case) =>
+            casesService.addAttachment(c, fileUpload, request.operator)
+              .flatMap(_ => successful(Redirect(routes.AttachmentsController.attachmentsDetails(reference))))
+          case _ =>
+            successful(Ok(views.html.case_not_found(reference)))
+        }
+      }
+      case _ => renderErrors(reference, Some("expected type file on the form"))
+    }
+  }
+
+  private def renderErrors(reference: String, errorMessage: Option[String])
```

Comment:
do we ever call `renderErrors(.., None)`? Seems like if we were calling `renderErrors` then we will probably always have `errors`?